File: app/src/main/java/com/example/udemy_thecompleteandroid14developercourse_build100apps/TheJournalApp/AddJournalActivity.java
package com.example.udemy_thecompleteandroid14developercourse_build100apps.TheJournalApp;

import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.example.udemy_thecompleteandroid14developercourse_build100apps.R;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.Timestamp;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

import java.util.Date;

public class AddJournalActivity extends AppCompatActivity {

    // Widgets
    private Button saveButton;
    private ImageView addPhotoBtn;
    private ProgressBar progressBar;
    private EditText titleEditText;
    private EditText thoughtsEditText;
    private ImageView imageView;


    // Firebase (FireStore)
    private FirebaseFirestore db = FirebaseFirestore.getInstance();
    private CollectionReference collectionReference =
            db.collection("Journal");


    // Firebase (Storage)
    private StorageReference storageReference;


    // Firebase Auth : UserId and UserName
    private String currentUserId;
    private String currentUserName;
    private FirebaseAuth firebaseAuth;
    private FirebaseAuth.AuthStateListener authStateListener;
    private FirebaseUser user;

    // Using Activity Result Launcher
    ActivityResultLauncher<String> mTakePhoto;
    Uri imageUri;




    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_journal);

        progressBar = findViewById(R.id.postProgressBar);
        titleEditText = findViewById(R.id.postTitle);
        thoughtsEditText = findViewById(R.id.postDescription);
        imageView = findViewById(R.id.postImage);
        saveButton = findViewById(R.id.savePostButton);
        addPhotoBtn = findViewById(R.id.postCameraButton);


        progressBar.setVisibility(View.INVISIBLE);

        // Firebase Storage Reference
        storageReference = FirebaseStorage.getInstance()
                .getReference();

        // Auth
        firebaseAuth = FirebaseAuth.getInstance();

        // Getting the Current User
        if (user != null){
            currentUserId = user.getUid();
            currentUserName = user.getDisplayName();
        }

        saveButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                SaveJournal();
            }
        });


        mTakePhoto = registerForActivityResult(
                new ActivityResultContracts.GetContent(),
                new ActivityResultCallback<Uri>() {
                    @Override
                    public void onActivityResult(Uri result) {
                        // Showing the image
                        imageView.setImageURI(result);

                        // Get the image URI
                        imageUri = result;
                    }
                }
        );




        addPhotoBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Getting Image From The Gallery
                mTakePhoto.launch("image/*");
            }
        });




    }

    private void SaveJournal() {

        String title = titleEditText.getText().toString().trim();
        String thoughts = thoughtsEditText.getText().toString().trim();

        progressBar.setVisibility(View.VISIBLE);

        if (!TextUtils.isEmpty(title) && !TextUtils.isEmpty(thoughts)
                && imageUri !=null){


            // the saving path of the images in Firebase Storage:
            // ........./journal_images/my_image_202310071430.png
            final StorageReference filePath = storageReference.
                    child("journal_images")
                    .child("my_image_"+ Timestamp.now().getSeconds());
            Log.d("uploadError", imageUri.toString());
            // Uploading the image
            filePath.putFile(imageUri)
                    .addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
                        @Override
                        public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {
                            filePath.getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
                                @Override
                                public void onSuccess(Uri uri) {
                                    String imageUrl = uri.toString();

                                    // Creating a Journal Object
                                    Journal journal = new Journal();
                                    journal.setTitle(title);
                                    journal.setThoughts(thoughts);
                                    journal.setImageUrl(imageUrl);

                                    journal.setTimeAdded(new Timestamp(new Date()));
                                    journal.setUserName(currentUserName);
                                    journal.setUserId(currentUserId);

                                    Log.d("uploadError", journal.getTitle());
                                    collectionReference.add(journal)
                                            .addOnSuccessListener(new OnSuccessListener<DocumentReference>() {
                                                @Override
                                                public void onSuccess(DocumentReference documentReference) {
                                                    progressBar.setVisibility(View.INVISIBLE);

                                                    Intent i = new Intent(AddJournalActivity.this, JournalListActivity.class);

                                                    startActivity(i);
                                                    finish();
                                                }
                                            }).addOnFailureListener(new OnFailureListener() {
                                                @Override
                                                public void onFailure(@NonNull Exception e) {
                                                    Toast.makeText(AddJournalActivity.this,
                                                            "Failed : "+e.getMessage(), Toast.LENGTH_SHORT).show();
                                                }
                                            });

                                }
                            });
                        }
                    }).addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(@NonNull Exception e) {
                            progressBar.setVisibility(View.INVISIBLE);
                            Log.d("uploadError", imageUri.toString());
                            Toast.makeText(AddJournalActivity.this,
                                    "Failed!!! "+e.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });


        }else{
            progressBar.setVisibility(View.INVISIBLE);
        }


    }

    @Override
    protected void onStart() {
        super.onStart();
        user = firebaseAuth.getCurrentUser();
    }


}

File: app/src/main/java/com/example/udemy_thecompleteandroid14developercourse_build100apps/KotlinJournalApp/AddJournalActivity.kt
package com.example.udemy_thecompleteandroid14developercourse_build100apps.KotlinJournalApp

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.example.udemy_thecompleteandroid14developercourse_build100apps.KotlinJournalApp.models.Journal
import com.example.udemy_thecompleteandroid14developercourse_build100apps.KotlinJournalApp.models.JournalUser
import com.example.udemy_thecompleteandroid14developercourse_build100apps.R
import com.example.udemy_thecompleteandroid14developercourse_build100apps.databinding.ActivityAddJournal2Binding
import com.google.firebase.Firebase
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.auth
import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import java.util.Date

class AddJournalActivity : AppCompatActivity() {
    
    lateinit var binding: ActivityAddJournal2Binding
    var currentUserId:String = ""
    var currentUserName:String = ""
    
    lateinit var auth:FirebaseAuth
    lateinit var user:FirebaseUser
    
    var db:FirebaseFirestore = FirebaseFirestore.getInstance()
    lateinit var storageReference: StorageReference
    
    var collectionReference: CollectionReference = db.collection("Journal")
    lateinit var imageUri:Uri
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_journal2)
        binding = ActivityAddJournal2Binding.inflate(layoutInflater)
        
        storageReference = FirebaseStorage.getInstance().getReference()
        
        auth = Firebase.auth    
        binding.apply { 
            postProgressBar.visibility = View.VISIBLE
            if(JournalUser.instance != null){
//                currentUserId = JournalUser.instance!!.userId.toString()
//                currentUserName = JournalUser.instance!!.userName.toString()
                currentUserId = auth.currentUser?.uid.toString()
                currentUserName = auth.currentUser?.displayName.toString()
                postUserName.text = currentUserName
                
            }
            savePostButton.setOnClickListener{
                saveJournal()
            }
        }
        
        binding.postCameraButton.setOnClickListener{
            val intent = Intent(Intent.ACTION_GET_CONTENT)
            intent.setType("image/*")
            startActivityForResult(intent, 1)
            
        }
       
        
    }

    private fun saveJournal() {
        var title:String = binding.postTitle.text.toString().trim()
        var thoughts:String = binding.postDescription.text.toString().trim()
        
        binding.postProgressBar.visibility = View.VISIBLE
        
        if(title.isNotEmpty() && thoughts.isNotEmpty() && imageUri != null){
            val filePath:StorageReference = storageReference
                .child("journal_images")
                .child("my_image_"+ Timestamp.now().seconds)

            filePath.putFile(imageUri).addOnSuccessListener { 
                filePath.downloadUrl.addOnSuccessListener { 
                    var imageUri:String = it.toString()
                    
                    val timeStamp:Timestamp = Timestamp(Date())
                    
                    var journal:Journal = Journal(title, thoughts,imageUri, currentUserId ,timeStamp, currentUserName)
                    
                    collectionReference.add(journal).addOnSuccessListener { 
                        binding.postProgressBar.visibility = View.VISIBLE
                        var intent = Intent(this, KotlinJournalsActivity::class.java)
                        startActivity(intent)
                        finish()
                    }.addOnFailureListener{
                        binding.postProgressBar.visibility = View.INVISIBLE
                    }
                }
            }
        } else {
            binding.postProgressBar.visibility = View.VISIBLE
        }
        
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 1 && resultCode == RESULT_OK){
            if (data != null){
                imageUri = data.data!!
                binding.postImage.setImageURI(imageUri)
            }
        }
    }

    override fun onStart() {
        super.onStart()
        user = auth.currentUser!!
        
    }

    override fun onStop() {
        super.onStop()
        if (auth != null){
            
        }
    }
}